shared/middleware/helpers/bucket_helper.rb

```diff
@@ -175,43 +175,80 @@ def create_or_replace(encrypted_channel_id, filename, body, version = nil, abuse
+  # different client more recently wrote to this project, logs an event to
+  # firehose and halts with 409 Conflict.
+  #
+  # In some cases, S3 replication lag could cause the current_version not to
+  # even appear in the version list. In this case, do not log or halt.
+  #
+  # Clients displaying projects must obey the following rules:
+  # (1) When loading a project, read its latest version.
+  # (2) When saving a project for the first time, create a new version, but
+  #     do not replace any previous versions.
+  # (3) When saving a project subsequent times, create a new version, and
+  #     replace the "current version" (i.e. last known version that client wrote).
+  #
+  # Therefore, a project will only ever replace a version that it created, and
+  # we can say the client "owns" a particular version if that client created it.
+  def check_current_version(encrypted_channel_id, filename, current_version, should_replace, timestamp, tab_id, user_id)
+    return true unless filename == 'main.json' && @bucket == CDO.sources_s3_bucket && current_version
 
     owner_id, channel_id = storage_decrypt_channel_id(encrypted_channel_id)
     key = s3_path owner_id, channel_id, filename
 
-    # check current version id without pulling down the whole object.
-    current_version = s3.get_object_tagging(bucket: @bucket, key: key).version_id
-
-    return if version_to_replace == current_version
+    # This is an Array of ObjectVersions, defined in:
+    # https://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Types/ObjectVersion.html
+    versions = s3.list_object_versions(bucket: @bucket, prefix: key).versions
+
+    target_version_metadata = versions.find {|v| v.version_id == current_version}
+
+    error_type =
+      if should_replace
+        # If we are replacing the target version, then we "own" it and don't have
+        # to worry about other clients replacing it. We *do* have to worry about
+        # the target version not being visible yet due to S3's read-after-write
+        # eventual consistency though, so allow the target version to either be
+        # (absent or (present and latest)).
+        return true unless target_version_metadata && !target_version_metadata.is_latest
+        'reject-replacing-older-main-json'
+      else
+        # Since we are not replacing the target version, we can conclude:
+        # (1) the client just loaded the project for the first time, meaning
+        #     the client recently successfully read this version, so we aren't
+        #     too worried about S3 inconsistency.
+        # (2) this version is owned by a different client, therefore another
+        #     client may have already replaced it.
+        # Guard against this scenario by requiring that the target version be
+        # both present and latest, without worrying about S3 inconsistency.
+        return true unless !target_version_metadata || !target_version_metadata.is_latest
```

Comment:
Removing the double-negative in the conditional here would match how you phrased this logic in your comment better:
```rb
return true if target_version_metadata&.is_latest
```